File: app/lib/services/auth_service.dart
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;
import '../models/user.dart';
import '../config/app_config.dart';

class AuthService {
  static const Duration _timeout = Duration(seconds: 10);

  // Mock пользователь для dev режима
  User _getMockUser() {
    return User(
      id: 'dev_user_001',
      name: 'Тестовый Пользователь',
      email: 'test@example.com',
      appleId: 'dev_apple_id_001',
      subscription: Subscription(
        plan: 'free',
        expiresAt: null,
        autoRenew: false,
      ),
      preferences: UserPreferences(
        notifications: true,
        theme: 'system',
        favoriteETFs: ['ARKB', 'IBIT'],
      ),
      createdAt: DateTime.now().subtract(const Duration(days: 30)),
      lastLoginAt: DateTime.now(),
    );
  }

  // Выход из аккаунта
  Future<void> signOut() async {
    try {
      final response = await http
          .post(Uri.parse(AppConfig.getApiUrl('/auth/logout')))
          .timeout(_timeout);

      if (response.statusCode != 200) {
        throw Exception('Ошибка выхода: ${response.statusCode}');
      }
    } catch (e) {
      if (e is TimeoutException) {
        throw Exception('Превышено время ожидания ответа от сервера');
      }
      throw Exception('Ошибка выхода: $e');
    }
  }

  // Проверка валидности токена
  Future<User> validateToken(String token) async {
    try {
      // Проверяем, находимся ли мы в dev режиме
      if (AppConfig.isDebugMode) {
        print('🔧 Dev режим: Используем mock данные для validateToken');
        return _getMockUser();
      }

      final response = await http
          .get(
            Uri.parse(AppConfig.getApiUrl('/auth/validate')),
            headers: {'Authorization': 'Bearer $token'},
          )
          .timeout(_timeout);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return User.fromJson(data['user']);
      } else {
        throw Exception('Недействительный токен: ${response.statusCode}');
      }
    } catch (e) {
      if (e is TimeoutException) {
        throw Exception('Превышено время ожидания ответа от сервера');
      }
      throw Exception('Ошибка проверки токена: $e');
    }
  }

  // Обновление профиля пользователя
  Future<User> updateProfile({
    required String userId,
    String? name,
    UserPreferences? preferences,
  }) async {
    try {
      final response = await http
          .put(
            Uri.parse(AppConfig.getApiUrl('/auth/profile')),
            headers: {'Content-Type': 'application/json'},
            body: json.encode({
              'userId': userId,
              if (name != null) 'name': name,
              if (preferences != null) 'preferences': preferences.toJson(),
            }),
          )
          .timeout(_timeout);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return User.fromJson(data['user']);
      } else {
        throw Exception('Ошибка обновления профиля: ${response.statusCode}');
      }
    } catch (e) {
      if (e is TimeoutException) {
        throw Exception('Превышено время ожидания ответа от сервера');
      }
      throw Exception('Ошибка обновления профиля: $e');
    }
  }

  // Обновление подписки
  Future<User> updateSubscription({
    required String userId,
    required Subscription subscription,
  }) async {
    try {
      final response = await http
          .put(
            Uri.parse(AppConfig.getApiUrl('/auth/subscription')),
            headers: {'Content-Type': 'application/json'},
            body: json.encode({
              'userId': userId,
              'subscription': subscription.toJson(),
            }),
          )
          .timeout(_timeout);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return User.fromJson(data['user']);
      } else {
        throw Exception('Ошибка обновления подписки: ${response.statusCode}');
      }
    } catch (e) {
      if (e is TimeoutException) {
        throw Exception('Превышено время ожидания ответа от сервера');
      }
      throw Exception('Ошибка обновления подписки: $e');
    }
  }

  // Регистрация устройства для push-уведомлений
  Future<void> registerDevice({
    required String userId,
    required String deviceToken,
  }) async {
    try {
      final response = await http
          .post(
            Uri.parse(AppConfig.getApiUrl('/auth/device')),
            headers: {'Content-Type': 'application/json'},
            body: json.encode({
              'userId': userId,
              'deviceToken': deviceToken,
              'platform': 'ios', // или 'android'
            }),
          )
          .timeout(_timeout);

      if (response.statusCode < 200 || response.statusCode >= 300) {
        throw Exception(
          'Ошибка регистрации устройства: ${response.statusCode}',
        );
      }
    } catch (e) {
      if (e is TimeoutException) {
        throw Exception('Превышено время ожидания ответа от сервера');
      }
      throw Exception('Ошибка регистрации устройства: $e');
    }
  }
}

File: backend/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { JwtService } from '@nestjs/jwt';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  async signInWithApple(appleData: {
    identityToken: string;
    authorizationCode: string;
    userIdentifier: string;
    givenName?: string;
    familyName?: string;
    email?: string;
  }) {
    try {
      // В реальном приложении здесь должна быть валидация токена Apple
      // Для демонстрации просто создаем или обновляем пользователя

      let user = await this.prisma.user.findUnique({
        where: { appleId: appleData.userIdentifier },
        include: {
          subscription: true,
          preferences: true,
        },
      });

      if (!user) {
        // Создаем нового пользователя
        user = await this.prisma.user.create({
          data: {
            appleId: appleData.userIdentifier,
            name:
              appleData.givenName && appleData.familyName
                ? `${appleData.givenName} ${appleData.familyName}`
                : appleData.givenName || appleData.familyName,
            email: appleData.email,
            lastLoginAt: new Date(),
            subscription: {
              create: {
                plan: 'free',
                autoRenew: false,
              },
            },
            preferences: {
              create: {
                notifications: true,
                theme: 'system',
                favoriteETFs: [],
              },
            },
          },
          include: {
            subscription: true,
            preferences: true,
          },
        });
      } else {
        // Обновляем время последнего входа
        user = await this.prisma.user.update({
          where: { id: user.id },
          data: { lastLoginAt: new Date() },
          include: {
            subscription: true,
            preferences: true,
          },
        });
      }

      // Генерируем JWT токен
      const token = this.generateToken(user.id);

      return {
        user: this.formatUserResponse(user),
        token,
      };
    } catch (error) {
      throw new UnauthorizedException('Ошибка аутентификации через Apple');
    }
  }

  async validateToken(token: string) {
    try {
      const payload = this.jwtService.verify(token);
      const user = await this.prisma.user.findUnique({
        where: { id: payload.sub },
        include: {
          subscription: true,
          preferences: true,
        },
      });

      if (!user) {
        throw new UnauthorizedException('Пользователь не найден');
      }

      return this.formatUserResponse(user);
    } catch (error) {
      throw new UnauthorizedException('Недействительный токен');
    }
  }

  async updateProfile(
    userId: string,
    data: {
      name?: string;
      preferences?: {
        notifications?: boolean;
        theme?: string;
        favoriteETFs?: string[];
      };
    },
  ) {
    const user = await this.prisma.user.update({
      where: { id: userId },
      data: {
        ...(data.name && { name: data.name }),
        ...(data.preferences && {
          preferences: {
            update: {
              notifications: data.preferences.notifications,
              theme: data.preferences.theme,
              favoriteETFs: data.preferences.favoriteETFs,
            },
          },
        }),
      },
      include: {
        subscription: true,
        preferences: true,
      },
    });

    return this.formatUserResponse(user);
  }

  async updateSubscription(
    userId: string,
    data: {
      plan: string;
      expiresAt?: Date;
      autoRenew?: boolean;
    },
  ) {
    const user = await this.prisma.user.update({
      where: { id: userId },
      data: {
        subscription: {
          upsert: {
            create: {
              plan: data.plan,
              expiresAt: data.expiresAt,
              autoRenew: data.autoRenew || false,
            },
            update: {
              plan: data.plan,
              expiresAt: data.expiresAt,
              autoRenew: data.autoRenew,
            },
          },
        },
      },
      include: {
        subscription: true,
        preferences: true,
      },
    });

    return this.formatUserResponse(user);
  }

  async registerDevice(userId: string, deviceToken: string, platform: string) {
    // В реальном приложении здесь можно сохранить токен устройства
    // для отправки push-уведомлений
    return { success: true };
  }

  private generateToken(userId: string): string {
    const payload = { sub: userId };
    return this.jwtService.sign(payload);
  }

  private formatUserResponse(user: any) {
    return {
      id: user.id,
      name: user.name,
      email: user.email,
      appleId: user.appleId,
      subscription: user.subscription
        ? {
            plan: user.subscription.plan,
            expiresAt: user.subscription.expiresAt,
            autoRenew: user.subscription.autoRenew,
          }
        : null,
      preferences: user.preferences
        ? {
            notifications: user.preferences.notifications,
            theme: user.preferences.theme,
            favoriteETFs: user.preferences.favoriteETFs,
          }
        : null,
      createdAt: user.createdAt,
      lastLoginAt: user.lastLoginAt,
    };
  }
}
